splash/qtutils.py

```diff
@@ -254,171 +256,380 @@ def swap_byte_order_i32(buf):
+            str(scale_method))
+
+
+def _render_qwebpage_vector(web_page, logger,
+                            in_viewport, out_viewport, image_size):
+    """
+    Render a webpage using vector rescale method.
+
+    :param in_viewport: region of the webpage to render from
+    :param out_viewport: region of the image to render to
+    :param image_size: size of the resulting image
+
+    :type in_viewport: QRect
+    :type out_viewport: QRect
+    :type image_size: QSize
+
+    """
+    web_rect = QRect(in_viewport)
+    render_rect = QRect(out_viewport)
+    canvas_size = QSize(image_size)
+
+    logger.log("png render: rendering %s of the web page" % web_rect,
+               min_level=2)
+    logger.log("png render: rendering into %s of the canvas" % render_rect,
+               min_level=2)
+    logger.log("png render: canvas size=%s" % canvas_size, min_level=2)
+
+    to_paint = render_rect.intersected(QRect(QPoint(0, 0), canvas_size))
+    # QPainter has several issues when its rendering area is more than 2**15 in
+    # any dimension.
+    need_tiling = max(to_paint.width(), to_paint.height()) >= (1 << 15)
+    if need_tiling:
+        logger.log("png render: draw region too large, rendering tile-by-tile",
+                   min_level=2)
+        return _render_qwebpage_tiled(web_page, logger,
+                                      web_rect, render_rect, canvas_size)
+    else:
+        logger.log("png render: rendering webpage in one step", min_level=2)
+        return _render_qwebpage_full(web_page, logger,
+                                     web_rect, render_rect, canvas_size)
+
+
+def _render_qwebpage_raster(web_page, logger,
+                            in_viewport, out_viewport, image_size):
+    """
+    Render a webpage rescaling pixel-wise if necessary.
+
+    :param in_viewport: region of the webpage to render from
+    :param out_viewport: region of the image to render to
+    :param image_size: size of the resulting image
+
+    :type in_viewport: QRect
+    :type out_viewport: QRect
+    :type image_size: QSize
+
+    """
+    render_rect = QRect(in_viewport)
+    if in_viewport.size() == out_viewport.size():
+        # If no resizing is requested, we can make canvas of the size of the
+        # output image to avoid the final cropping step.
+        canvas_size = QSize(image_size)
+    else:
+        canvas_size = QSize(in_viewport.size())
+
+    if image_size.height() < out_viewport.height():
+        logger.log("png render: image is trimmed vertically")
+        hcut = image_size.height()
+        if in_viewport.size() == out_viewport.size():
+            # If no resizing then we need to render exactly the requested
+            # number of pixels vertically.
+            hrender = hcut
+        else:
+            # If there's a resize, there will be interpolation.  We must ensure
+            # that both pixels that contribute to the color of the last row of
+            # the resulting image are rendered.
```

Comment:
No more than 2 pixels in any dimension contribute to the value of an output pixel in bilinear interpolation algorithm. Bicubic or Lanczos filters have bigger windows, so we'll need to update the formula accordingly before using one of those.